csharp/examples/IoTExample/Model/SampleDataPipeline.cs

```diff
@@ -0,0 +1,197 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.IO;
+using System.Threading.Tasks;
+using System.Collections.Generic;
+using Apache.Arrow;
+using Apache.Arrow.Ipc;
+using Apache.Arrow.Memory;
+using System.Threading.Channels;
+using System.Threading;
+using System.Collections.Concurrent;
+
+namespace IoTPipelineExample
+{
+    public class SampleDataPipeline
+    {
+        private int _size;
+        private int _partitions;
+        private readonly int _inputs;
+        private readonly int _capacity;
+        private readonly Channel<SensorData> _channel;
+        ChannelWriter<SensorData> _writer;
+        ChannelReader<SensorData> _reader;
+
+        private readonly List<ConcurrentBag<int>> _colSubjectIdArrays;
+        private readonly List<ConcurrentBag<string>> _colActivityLabelArrays;
+        private readonly List<ConcurrentBag<long>> _colTimestampArrays;
+        private readonly List<ConcurrentBag<double>> _colXAxisArrays;
+        private readonly List<ConcurrentBag<double>> _colYAxisArrays;
+        private readonly List<ConcurrentBag<double>> _colZAxisArrays;
+
+        public Dictionary<string, string> activityLabel = new Dictionary<string, string>()
+            {
+                {"walking", "A"},
+                {"jogging", "B"},
+                {"stairs", "C"},
+                {"sitting", "D"},
+                {"standing", "E"},
+                {"typing", "F"},
+                {"teeth", "G"},
+                {"soup", "H"},
+                {"chips", "I"},
+                {"pasta", "J"},
+                {"drinking", "K"},
+                {"sandwich", "L"},
+                {"kicking", "M"},
+                {"catch", "O"},
+                {"dribbling", "P"},
+                {"writing", "Q"},
+                {"clapping", "R"},
+                {"folding", "S"},
+            };
+
+        public SampleDataPipeline(int concurrencyLevel, int totalSensorData, int queueCapacity)
+        {
+            _partitions = concurrencyLevel;
+            _inputs = totalSensorData;
+            _capacity = queueCapacity;
+            _channel = Channel.CreateBounded<SensorData>(_capacity);
+            _writer = _channel.Writer;
+            _reader = _channel.Reader;
+
+            _colSubjectIdArrays = new List<ConcurrentBag<int>>();
+            _colActivityLabelArrays = new List<ConcurrentBag<string>>();
+            _colTimestampArrays = new List<ConcurrentBag<long>>();
+            _colXAxisArrays = new List<ConcurrentBag<double>>();
+            _colYAxisArrays = new List<ConcurrentBag<double>>();
+            _colZAxisArrays = new List<ConcurrentBag<double>>();
+
+            for (int i = 0; i < _partitions; i++)
+            {
+                _colSubjectIdArrays.Add(new ConcurrentBag<int>());
```

Comment:
ConcurrentBag is unordered. Meaning when you enumerate it, you can get the results in a different order than what they went in. Is that intended?

https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentbag-1?view=net-6.0

Reply:
My intention was to find a data structure that has below characteristics:

1. Inserting by row
2. Querying by column
3. The "results in a different order than what they went in" does not matter, but the kth item in one column should belong to the same row as the kth item in another column
4. So all of the column arrays can be appended to a RecordBatch

It seems DataTable is a good candidate, or could you recommend other light weight data structures instead?

Reply:
> but the kth item in one column should belong to the same row as the kth item in another column

Since a "bag" is unordered, there is no such thing as the "kth item". Notice there is no indexer. You can enumerate the items, but you aren't guaranteed the order they are returned.

> It seems DataTable is a good candidate, or could you recommend other light weight data structures instead?

I wouldn't use DataTable. Instead, I would eliminate the concurrency in `ReadFromChannel`. Have many "writer" threads all writing items to the channel. But then only have 1 "reader" thread that pulls items off the channel and writes them directly to a `RecordBatch.Builder`.

Reply:
I have refactored the code per your suggestion to use multiple writers and 1 reader only, and writes data directly to builder.
And I also modified the example to emulate below scenario:

- There are around 9,000 sensors(subjects) in total
- Each sensor can send hundreds or thousands events during a short time of period
- When reading from channel, data is appended to each array builder hashed by SubjectId as the key of builderId
- When building RecordBatch, the array builders with the same builderId are built together
- The final Arrow file contains 9,000 RecordBatches 

Do you think this a good scenario for real world applications in terms of utilizing Arrow RecordBatch?